File: frontend/src/types/meeting.ts
export interface Meeting {
  PK: string;
  SK: string;
  meetingId: string;
  userId: string;
  title: string;
  description?: string;
  date: string;
  duration?: number;
  status: 'recording' | 'transcribing' | 'summarizing' | 'done' | 'error';
  tags?: string[];
  sentiment?: 'positive' | 'neutral' | 'negative';
  participants?: Participant[];
  summary?: string;
  actionItems?: ActionItem[];
  transcription?: TranscriptSegment[];
  attachments?: Attachment[];
  sharedWith?: SharedUser[];
  createdAt: string;
  updatedAt: string;
  sttProvider?: 'transcribe' | 'nova-sonic' | 'whisper';
  /** Per ADR-014: ordered S3 keys for multi-file uploads. Falls back to audioKey for legacy single-file meetings. */
  audioKey?: string;
  audioKeys?: string[];
  /** Per ADR-014 Phase 6: ordered predecessor meeting IDs whose summaries are prepended to this meeting's prompt. */
  linkedMeetingIds?: string[];
  /** ADR-031 cost/sizing simulator, singleton per meeting. */
  simRun?: SimRun;
}

/** Mirrors backend/internal/model.SimRequirement's JSON shape exactly. */
export interface SimRequirement {
  key: string;
  label: string;
  value: string;
  unit?: string;
  required: boolean;
  source: 'extracted' | 'user';
  /** transcript://{segmentId} deep link (ADR-013), empty when user-entered. */
  evidence?: string;
}

export interface SimOption {
  name: string;
  description?: string;
}

export interface SimChart {
  key: string;
  url?: string;
}

/** Mirrors backend/internal/model.SimRunResponse's JSON shape (ADR-031). */
export interface SimRun {
  simRunId: string;
  status: 'extracted' | 'queued' | 'running' | 'done' | 'error';
  requirements?: SimRequirement[];
  options?: SimOption[];
  charts?: SimChart[];
  reportMarkdown?: string;
  codeKey?: string;
  priceSnapshotAt?: string;
  errorMessage?: string;
  createdAt: string;
  updatedAt: string;
}

export interface Participant {
  id: string;
  name: string;
  email?: string;
  avatarUrl?: string;
  initials?: string;
}

export interface ActionItem {
  id: string;
  text: string;
  completed: boolean;
  assignee?: string;
  dueDate?: string;
}

export interface TranscriptSegment {
  id: string;
  speaker: string;
  speakerInitials?: string;
  timestamp: string;
  text: string;
  startTime: number;
  endTime: number;
}

export interface Attachment {
  id: string;
  name: string;
  type: 'image' | 'document' | 'audio' | 'video' | 'photo' | 'screenshot' | 'diagram' | 'whiteboard' | 'audio_file';
  url: string;
  thumbnailUrl?: string;
  originalUrl?: string;
  processedUrl?: string;
  processedContent?: string;
  size?: number;
  mimeType?: string;
  status?: string;
  timestamp?: string;
  createdAt: string;
}

export interface SharedUser {
  userId: string;
  email: string;
  name?: string;
  permission: 'read' | 'edit';
  sharedAt: string;
}

export interface TranscriptComparison {
  meetingId: string;
  providerA: {
    name: string;
    segments: TranscriptSegment[];
  };
  providerB: {
    name: string;
    segments: TranscriptSegment[];
  };
}

export interface MeetingListFilter {
  tab: 'all' | 'recent' | 'shared' | 'favorites';
  search?: string;
  tags?: string[];
  dateRange?: {
    start: string;
    end: string;
  };
}

export interface User {
  userId: string;
  email: string;
  name?: string;
  avatarUrl?: string;
}

// Extended meeting detail from API
export interface MeetingDetail extends Meeting {
  content?: string;
  notes?: string;
  /** Real-time summary built during recording (markdown incl. mermaid) */
  liveSummary?: string;
  transcriptA?: string;
  transcriptB?: string;
  selectedTranscript?: 'A' | 'B' | null;
  audioKey?: string;
  audioKeys?: string[];
  audioPartCount?: number;
  audioPartsReady?: number;
  speakerMap?: Record<string, string>;
  shares?: SharedUser[];
  isShared?: boolean;
  sharedBy?: string | null;
  permission?: 'read' | 'edit' | null;
  accountId?: string;
  sharedToAccount?: boolean;
}

export interface KBFile {
  fileId: string;
  fileName: string;
  fileType?: string;
  size?: number;
  lastModified: string;
}

export interface QAEntry {
  id: string;
  question: string;
  answer: string;
  sources?: string[];
  usedKB?: boolean;
  usedDocs?: boolean;
  toolsUsed?: string[];
  timestamp: string;
}

export interface IntegrationConfig {
  configured: boolean;
  maskedKey?: string;
  parentPageId?: string; // empty on a legacy record saved before the parent-page requirement — needs re-connect
}

export interface IntegrationsResponse {
  notion?: IntegrationConfig;
}

export interface ExportResponse {
  url?: string;
  notionPageId?: string;
  notionUrl?: string;
  filename?: string;
  content?: string;
}

export interface CrawlerSource {
  sourceId: string;
  sourceName: string;
  subscribers: string[];
  awsServices: string[];
  newsQueries: string[];
  newsSources: string[];
  customUrls: string[];
  schedule: string;
  lastCrawledAt: string;
  status: string;
  documentCount: number;
}

export interface CrawlerSubscription {
  sourceId: string;
  awsServices: string[];
  newsSources: string[];
  customUrls: string[];
  addedAt: string;
}

export interface CrawlerSourceResponse {
  source: CrawlerSource;
  subscription: CrawlerSubscription;
}

export interface CrawledDocument {
  docHash: string;
  sourceId?: string;
  type: 'news' | 'tech';
  title: string;
  url: string;
  source?: string;
  summary?: string;
  awsServices?: string[];
  tags?: string[];
  s3Key?: string;
  crawledAt: number | string;
  inKB?: boolean;
  pubDate?: string;
}

export interface CrawlHistory {
  timestamp: string;
  docsAdded: number;
  docsUpdated: number;
  errors: string[];
  duration: number;
}

export interface ChatMessage {
  msgId: string;
  role: 'user' | 'agent';
  content: string;
  action?: 'propose_structure' | 'ask_question' | 'approve' | 'request_subpage' | 'respond';
  metadata?: string;
  createdAt: string;
}

export interface Research {
  researchId: string;
  userId?: string;
  /** User-editable display label, defaults to `topic` at creation. `topic` (the original prompt) stays immutable. */
  title?: string;
  topic: string;
  mode: 'quick' | 'standard' | 'deep';
  status: 'planning' | 'approved' | 'running' | 'done' | 'error';
  parentId?: string;
  createdAt: string;
  completedAt?: string;
  s3Key?: string;
  sourceCount?: number;
  wordCount?: number;
  summary?: string;
  errorMessage?: string;
  trashedAt?: string;
  isShared?: boolean;
  sharedBy?: string;
  accountIds?: string[];
}

export interface ResearchDetail extends Research {
  content?: string;
  shares?: SharedUser[];
}

export interface AccountResearchRef {
  researchId: string;
  topic: string;
  summary?: string;
  status: Research['status'];
  ownerUserId: string;
  createdAt: string;
}

// Assignable account team roles (owner is server-assigned, never offered here).
// Keep in sync with backend/internal/model/account.go's AssignableRoles.
export const ASSIGNABLE_ACCOUNT_ROLES = ['AM', 'TAM', 'SSA', 'SA', 'SA Manager', 'AM Manager'] as const;

export interface AccountSummary {
  accountId: string;
  name: string;
  role: string;
}

export interface AccountMember {
  // Omitted (not empty string) on a pending grant -- the backend DTO uses
  // `json:"userId,omitempty"`.
  userId?: string;
  email?: string;
  role: string;
  // true when the invitee has an invited-but-not-yet-logged-in Cognito
  // account and no DynamoDB profile exists yet -- the grant is queued and
  // becomes a real membership automatically on their first login. userId
  // is omitted in that case.
  pending?: boolean;
}

export interface Account {
  accountId: string;
  name: string;
  aliases?: string[];
  domains?: string[];
  industry?: string;
  ownerUserId: string;
  members: AccountMember[];
  createdAt: string;
}

export interface AccountMeetingRef {
  meetingId: string;
  ownerUserId: string;
  title: string;
  date: string;
}

export interface FieldInsight {
  type: string;
  text: string;
  evidence?: string;
  implication?: string;
  nextAction?: string;
  sourceType?: string;
  sourceId: string;
  occurredAt: string;
  tsMarker?: string;
  entities?: string[];
}

export interface AccountInsight extends FieldInsight {
  sourceType: string;
}

export interface ProjectMember {
  userId: string;
  email?: string;
}

export interface ProjectSummary {
  projectId: string;
  name: string;
  stage?: string;
  sfdcOpptyId?: string;
}

export interface Project {
  projectId: string;
  name: string;
  description?: string;
  sfdcOpptyId?: string;
  sfdcUrl?: string;
  stage?: string;
  ownerUserId: string;
  accountIds: string[];
  members: ProjectMember[];
  createdAt: string;
  updatedAt: string;
}

export interface ProjectMeetingRef {
  meetingId: string;
  ownerUserId: string;
  title: string;
  date: string;
}

export interface ProjectResearchRef {
  researchId: string;
  topic: string;
  summary?: string;
  status: string;
  ownerUserId: string;
  createdAt: string;
}

export type ProjectInsight = FieldInsight;

export interface ProjectBrief {
  project: Project;
  insightsByType: Record<string, ProjectInsight[]>;
  meetings: ProjectMeetingRef[];
  research: ProjectResearchRef[];
}

export interface AccountDocument {
  docId: string;
  title: string;
  docType?: string;
  path?: string;
  links?: string[];
  fileName?: string;
  sourceUserId: string;
  createdAt: string;
  updatedAt?: string;
  content?: string;
  downloadUrl?: string;
  /** Presigned URL for the PDF sidecar of a PPTX/PPT slide, once the
   * convert-doc Lambda's conversion has finished. Absent otherwise. */
  previewUrl?: string;
  /** Set once a public share link has been minted for this (personal,
   * slide) document; see docApi.createPublicShare. */
  publicShareToken?: string;
  /** Owner's email — present only on a document someone shared directly with
   * the current user. Read-only: edit/delete/share stay with the owner. */
  sharedBy?: string;
}

export interface PutDocumentRequest {
  title: string;
  docType?: string;
  path?: string;
  markdown?: string;
  fileKey?: string;
  fileName?: string;
  mimeType?: string;
  fileSize?: number;
}

export const INSIGHT_TYPES = [
  'trend', 'need', 'competitive', 'risk', 'opportunity', 'tech', 'stakeholder', 'action',
] as const;

export interface DictionaryTerm {
  phrase: string;
  soundsLike: string;
  displayAs: string;
}

export interface ChatSession {
  sessionId: string;
  title: string;
  createdAt: string;
  lastMessageAt: string;
  messageCount: number;
}

export interface DictionaryTerm {
  phrase: string;
  soundsLike: string;
  displayAs: string;
}

File: backend/internal/model/meeting.go
package model

import "time"

// Meeting represents a meeting record in DynamoDB
// PK: USER#{userId}, SK: MEETING#{meetingId}
type Meeting struct {
	PK                 string    `dynamodbav:"PK"`
	SK                 string    `dynamodbav:"SK"`
	MeetingID          string    `dynamodbav:"meetingId"`
	UserID             string    `dynamodbav:"userId"`
	Title              string    `dynamodbav:"title"`
	Date               time.Time `dynamodbav:"date"`
	Content            string    `dynamodbav:"content,omitempty"`                      // Markdown meeting notes
	TranscriptA        string    `dynamodbav:"transcriptA,omitempty"`                  // AWS Transcribe result
	TranscriptB        string    `dynamodbav:"transcriptB,omitempty"`                  // Nova Sonic result
	SelectedTranscript string    `dynamodbav:"selectedTranscript,omitempty"`           // "A" or "B"
	AudioKey           string    `dynamodbav:"audioKey,omitempty"`                     // S3 key for audio file (legacy single-file)
	AudioKeys          []string  `dynamodbav:"audioKeys,omitempty"`                    // Ordered S3 keys for multi-file uploads
	AudioPartCount     int       `dynamodbav:"audioPartCount,omitempty"`               // Total parts expected
	AudioPartsReady    int       `dynamodbav:"audioPartsReady,omitempty"`              // Parts with completed transcription
	AudioPartsReadySet []int     `dynamodbav:"audioPartsReadySet,omitempty,numberset"` // Set of completed part indices (source-of-truth for idempotent counting)
	// AllPartsEmittedAt marks the moment we first emitted the
	// `AllPartsTranscribed` EventBridge event for this meeting. Used as a
	// once-only lock so EventBridge at-least-once re-deliveries of part
	// transcripts don't republish duplicate events. Populated via
	// conditional update `SET allPartsEmittedAt = :now WHERE attribute_not_exists(...)`.
	AllPartsEmittedAt  string            `dynamodbav:"allPartsEmittedAt,omitempty"`
	SttProvider        string            `dynamodbav:"sttProvider,omitempty"`        // "transcribe" or "nova-sonic"
	TranscriptSegments string            `dynamodbav:"transcriptSegments,omitempty"` // JSON string of speaker-labeled segments
	ActionItems        string            `dynamodbav:"actionItems,omitempty"`        // JSON string of extracted action items
	Notes              string            `dynamodbav:"notes,omitempty"`              // User-written meeting notes (post-recording)
	LiveSummary        string            `dynamodbav:"liveSummary,omitempty"`        // Real-time summary built during recording (markdown incl. mermaid)
	SpeakerMap         map[string]string `dynamodbav:"speakerMap,omitempty"`         // spk_0 -> "김팀장" mapping
	// DiarizationSpeakerHint overrides len(Participants) as pyannote's
	// max_speakers bound when re-diarizing a meeting on demand (see
	// RediarizeMeeting) — set once a user-supplied headcount is known to be
	// more accurate than the registered Participants list.
	DiarizationSpeakerHint int      `dynamodbav:"diarizationSpeakerHint,omitempty"`
	Participants           []string `dynamodbav:"participants,omitempty"`
	Tags                   []string `dynamodbav:"tags,omitempty"`
	Sentiment              string   `dynamodbav:"sentiment,omitempty"`        // "positive", "neutral", "negative" — extracted by summarize Lambda
	Duration               int      `dynamodbav:"duration,omitempty"`         // Total audio length in seconds, written by the summarize Lambda
	Insights               string   `dynamodbav:"insights,omitempty"`         // JSON []MeetingInsight
	LinkedMeetingIDs       []string `dynamodbav:"linkedMeetingIds,omitempty"` // Chronologically ordered predecessor IDs
	AccountID              string   `dynamodbav:"accountId,omitempty"`        // linked Account
	// ProjectIDs links this meeting to zero or more projects (many-to-many).
	// Stored as a DynamoDB String Set (stringset tag), not a List, so link
	// mutations can use atomic ADD/DELETE without a read-modify-write race.
	ProjectIDs      []string  `dynamodbav:"projectIds,omitempty,stringset" json:"projectIds,omitempty"`
	SharedToAccount bool      `dynamodbav:"sharedToAccount,omitempty"` // published to account team
	NotionPageID    string    `dynamodbav:"notionPageId,omitempty"`    // Notion page created by a prior export; re-export updates it in place instead of creating a duplicate
	Status          string    `dynamodbav:"status"`                    // recording, transcribing, summarizing, done, error
	CreatedAt       time.Time `dynamodbav:"createdAt"`
	UpdatedAt       time.Time `dynamodbav:"updatedAt"`
	GSI1PK          string    `dynamodbav:"GSI1PK,omitempty"` // USER#{userId} for date sorting
	GSI1SK          string    `dynamodbav:"GSI1SK,omitempty"` // timestamp for sorting
	EntityType      string    `dynamodbav:"entityType"`       // "MEETING"
}

// Attachment represents a file attachment for a meeting
// PK: MEETING#{meetingId}, SK: ATTACH#{attachmentId}
type Attachment struct {
	PK               string    `dynamodbav:"PK"`
	SK               string    `dynamodbav:"SK"`
	AttachmentID     string    `dynamodbav:"attachmentId"`
	MeetingID        string    `dynamodbav:"meetingId"`
	UserID           string    `dynamodbav:"userId"`
	OriginalKey      string    `dynamodbav:"originalKey"`
	ProcessedKey     string    `dynamodbav:"processedKey,omitempty"`
	Type             string    `dynamodbav:"type"`   // photo, screenshot, diagram, whiteboard
	Status           string    `dynamodbav:"status"` // uploaded, processing, done
	Description      string    `dynamodbav:"description,omitempty"`
	ProcessedContent string    `dynamodbav:"processedContent,omitempty"` // Mermaid/markdown result
	FileName         string    `dynamodbav:"fileName,omitempty"`
	FileSize         int64     `dynamodbav:"fileSize,omitempty"`
	MimeType         string    `dynamodbav:"mimeType,omitempty"`
	CreatedAt        time.Time `dynamodbav:"createdAt"`
	EntityType       string    `dynamodbav:"entityType"` // "ATTACHMENT"
}

// Share represents a shared meeting access record
// For recipient lookup: PK: USER#{sharedToUserId}, SK: SHARED#{meetingId}
// For meeting's share list: PK: MEETING#{meetingId}, SK: SHARE_TO#{userId}
type Share struct {
	PK         string `dynamodbav:"PK"`
	SK         string `dynamodbav:"SK"`
	MeetingID  string `dynamodbav:"meetingId"`
	OwnerID    string `dynamodbav:"ownerId"`
	OwnerEmail string `dynamodbav:"ownerEmail,omitempty"`
	SharedToID string `dynamodbav:"sharedToId"`
	Email      string `dynamodbav:"email"`
	Permission string `dynamodbav:"permission"` // "read" or "edit"
	// Origin distinguishes a share created by ShareMeetingToAccount ("account")
	// from one created by the owner directly via ShareMeetingByEmail (""/direct).
	// RemoveMember's cleanup must only ever delete "account"-origin shares -- a
	// direct share is a separate grant the owner made explicitly and removing a
	// team member must never silently revoke it.
	Origin string `dynamodbav:"origin,omitempty"`
	// AccountID is set only on Origin=="account" shares (by CreateShareIfMember,
	// and retroactively by the backfill CLI) to the account that granted them.
	// DeleteShareIfAccountOrigin ties its delete condition to this field so a
	// meeting re-shared to a DIFFERENT account between RemoveMember's cleanup
	// read and its delete can never have that new account's grant swept up by
	// the old account's removal -- the meeting.AccountID re-check that guards
	// the read is a separate, non-atomic call and can't by itself prevent that
	// race; only a condition on the row being deleted can.
	AccountID  string    `dynamodbav:"accountId,omitempty"`
	CreatedAt  time.Time `dynamodbav:"createdAt"`
	EntityType string    `dynamodbav:"entityType"` // "SHARE"
}

const (
	ShareOriginAccount = "account"
)

// User represents a user record (for search functionality)
// PK: USER#{userId}, SK: PROFILE
type User struct {
	PK         string    `dynamodbav:"PK"`
	SK         string    `dynamodbav:"SK"`
	UserID     string    `dynamodbav:"userId"`
	Email      string    `dynamodbav:"email"`
	Name       string    `dynamodbav:"name,omitempty"`
	CreatedAt  time.Time `dynamodbav:"createdAt"`
	GSI2PK     string    `dynamodbav:"GSI2PK,omitempty"` // EMAIL#{email} for email search
	GSI2SK     string    `dynamodbav:"GSI2SK,omitempty"` // USER#{userId}
	EntityType string    `dynamodbav:"entityType"`       // "USER"
}

// UserLogin represents the last-login timestamp for the admin user-management
// panel, written by the Cognito PostAuthentication trigger
// (infra/lambda/post-authentication). Deliberately a separate item from
// User/PROFILE above: writing lastLoginAt onto PROFILE would risk creating a
// stub profile item (missing email/GSI2PK/GSI2SK) if the trigger's write
// raced GetOrCreateUser on a user's very first login, silently breaking
// email-based search/sharing for that user. This item carries no GSI keys.
// PK: USER#{userId}, SK: LOGIN
type UserLogin struct {
	PK          string    `dynamodbav:"PK"`
	SK          string    `dynamodbav:"SK"`
	LastLoginAt time.Time `dynamodbav:"lastLoginAt"`
	EntityType  string    `dynamodbav:"entityType"` // "USER_LOGIN"
}

// CrawlerSource represents a crawler source configuration
// PK: USER#{userId}, SK: CRAWLER#{sourceId}
type CrawlerSource struct {
	SourceID      string   `dynamodbav:"sourceId" json:"sourceId"`
	SourceName    string   `dynamodbav:"sourceName" json:"sourceName"`
	OwnerID       string   `dynamodbav:"ownerId,omitempty" json:"ownerId,omitempty"` // creator; subscribing later never changes this
	Subscribers   []string `dynamodbav:"subscribers" json:"subscribers"`
	AWSServices   []string `dynamodbav:"awsServices" json:"awsServices"`
	NewsQueries   []string `dynamodbav:"newsQueries" json:"newsQueries"`
	NewsSources   []string `dynamodbav:"newsSources" json:"newsSources"`
	CustomUrls    []string `dynamodbav:"customUrls" json:"customUrls"`
	Schedule      string   `dynamodbav:"schedule" json:"schedule"`
	LastCrawledAt string   `dynamodbav:"lastCrawledAt" json:"lastCrawledAt"`
	Status        string   `dynamodbav:"status" json:"status"`
	DocumentCount int      `dynamodbav:"documentCount" json:"documentCount"`
}

// CrawlerSubscription represents a user's subscription to a crawler source
// PK: USER#{userId}, SK: CRAWL_SUB#{sourceId}
type CrawlerSubscription struct {
	SourceID    string   `dynamodbav:"sourceId" json:"sourceId"`
	AWSServices []string `dynamodbav:"awsServices" json:"awsServices"`
	NewsSources []string `dynamodbav:"newsSources" json:"newsSources"`
	NewsQueries []string `dynamodbav:"newsQueries" json:"newsQueries"`
	CustomUrls  []string `dynamodbav:"customUrls" json:"customUrls"`
	AddedAt     string   `dynamodbav:"addedAt" json:"addedAt"`
}

// CrawledDocument represents a document fetched by the crawler
// PK: CRAWLER#{sourceId}, SK: DOC#{docHash}
type CrawledDocument struct {
	DocHash     string   `dynamodbav:"docHash,omitempty" json:"docHash"`
	SourceID    string   `dynamodbav:"-" json:"sourceId,omitempty"`
	Type        string   `dynamodbav:"type" json:"type"`
	Title       string   `dynamodbav:"title" json:"title"`
	URL         string   `dynamodbav:"url" json:"url"`
	Source      string   `dynamodbav:"source,omitempty" json:"source"`
	Summary     string   `dynamodbav:"summary,omitempty" json:"summary"`
	AWSServices []string `dynamodbav:"awsServices,omitempty" json:"awsServices,omitempty"`
	Tags        []string `dynamodbav:"tags,omitempty" json:"tags,omitempty"`
	S3Key       string   `dynamodbav:"s3Key,omitempty" json:"s3Key"`
	CrawledAt   int64    `dynamodbav:"crawledAt" json:"crawledAt"`
	InKB        bool     `dynamodbav:"inKB,omitempty" json:"inKB"`
	PubDate     string   `dynamodbav:"pubDate,omitempty" json:"pubDate,omitempty"`
}

// CrawlHistory represents a crawl execution history entry
// PK: CRAWLER#{sourceId}, SK: HISTORY#{timestamp}
type CrawlHistory struct {
	Timestamp   string   `dynamodbav:"timestamp" json:"timestamp"`
	DocsAdded   int      `dynamodbav:"docsAdded" json:"docsAdded"`
	DocsUpdated int      `dynamodbav:"docsUpdated" json:"docsUpdated"`
	Errors      []string `dynamodbav:"errors" json:"errors"`
	Duration    int      `dynamodbav:"duration" json:"duration"`
}

// Research represents a deep research task
// PK: USER#{userId}, SK: RESEARCH#{researchId}
type Research struct {
	ResearchID string `dynamodbav:"researchId" json:"researchId"`
	UserID     string `dynamodbav:"userId" json:"userId"`
	// Title is a user-editable display label, defaulting to Topic at
	// creation. Kept distinct from Topic (the original research prompt,
	// permanently immutable) so a rename never mutates the prompt that
	// the agent pipeline (research-worker, research-agent) actually acts
	// on -- only the UI's display string changes.
	Title        string `dynamodbav:"title,omitempty" json:"title,omitempty"`
	Topic        string `dynamodbav:"topic" json:"topic"`
	Mode         string `dynamodbav:"mode" json:"mode"`
	Status       string `dynamodbav:"status" json:"status"`
	CreatedAt    string `dynamodbav:"createdAt" json:"createdAt"`
	CompletedAt  string `dynamodbav:"completedAt,omitempty" json:"completedAt,omitempty"`
	S3Key        string `dynamodbav:"s3Key,omitempty" json:"s3Key,omitempty"`
	SourceCount  int    `dynamodbav:"sourceCount,omitempty" json:"sourceCount,omitempty"`
	WordCount    int    `dynamodbav:"wordCount,omitempty" json:"wordCount,omitempty"`
	Summary      string `dynamodbav:"summary,omitempty" json:"summary,omitempty"`
	ErrorMessage string `dynamodbav:"errorMessage,omitempty" json:"errorMessage,omitempty"`
	ParentID     string `dynamodbav:"parentId,omitempty" json:"parentId,omitempty"`
	TrashedAt    string `dynamodbav:"trashedAt,omitempty" json:"trashedAt,omitempty"`
	IsShared     bool   `dynamodbav:"-" json:"isShared,omitempty"`
	SharedBy     string `dynamodbav:"-" json:"sharedBy,omitempty"`
	// AccountIDs links this research to zero or more accounts (many-to-many),
	// mirroring Meeting.AccountID but pluralized since one research report can
	// be relevant to several customer accounts. See account.go's ResearchRef
	// for the reverse-lookup item stored in each account's partition.
	//
	// Stored as a DynamoDB String Set (stringset tag), not a List -- this
	// lets LinkAccount/UnlinkAccount use atomic ADD/DELETE update
	// expressions instead of a read-modify-write on the whole list, which
	// would otherwise lose one side's change when two accounts are
	// (un)linked concurrently (ADD/DELETE on a set has no such race, and is
	// naturally idempotent for a value already present/absent).
	AccountIDs []string `dynamodbav:"accountIds,omitempty,stringset" json:"accountIds,omitempty"`
	// ProjectIDs links this research to zero or more projects (many-to-many).
	// Stored as a DynamoDB String Set (stringset tag), not a List, so link
	// mutations can use atomic ADD/DELETE without a read-modify-write race.
	ProjectIDs []string `dynamodbav:"projectIds,omitempty,stringset" json:"projectIds,omitempty"`
}

const MaxAudioParts = 10

func (m *Meeting) GetEffectiveAudioKeys() []string {
	if len(m.AudioKeys) > 0 {
		return m.AudioKeys
	}
	if m.AudioKey != "" {
		return []string{m.AudioKey}
	}
	return nil
}

func (m *Meeting) IsMultiPart() bool {
	return m.AudioPartCount > 1
}

// MeetingStatus constants
const (
	StatusRecording    = "recording"
	StatusTranscribing = "transcribing"
	StatusSummarizing  = "summarizing"
	StatusDone         = "done"
	StatusError        = "error"
)

// Permission constants
const (
	PermissionRead = "read"
	PermissionEdit = "edit"
)

// Attachment type constants (per API spec: photo, screenshot, diagram, whiteboard, document, video, audio_file)
const (
	AttachTypePhoto      = "photo"
	AttachTypeScreenshot = "screenshot"
	AttachTypeDiagram    = "diagram"
	AttachTypeWhiteboard = "whiteboard"
	AttachTypeDocument   = "document"
	AttachTypeVideo      = "video"
	AttachTypeAudioFile  = "audio_file"
)

// Attachment status constants
const (
	AttachStatusUploaded   = "uploaded"
	AttachStatusProcessing = "processing"
	AttachStatusDone       = "done"
)

// Key prefixes for single table design
const (
	PrefixUser       = "USER#"
	PrefixMeeting    = "MEETING#"
	PrefixAttachment = "ATTACH#"
	PrefixShare      = "SHARED#"
	PrefixShareTo    = "SHARE_TO#"
	PrefixEmail      = "EMAIL#"
	PrefixProfile    = "PROFILE"
	PrefixCrawler    = "CRAWLER#"
	PrefixCrawlSub   = "CRAWL_SUB#"
	PrefixDoc        = "DOC#"
	PrefixHistory    = "HISTORY#"
	PrefixConfig     = "CONFIG"
	PrefixResearch   = "RESEARCH#"

	// PrefixPendingShare / PrefixPendingAccount / PrefixPendingMeeting key the
	// PendingShare item below: PK: PENDINGSHARE#{email}, SK:
	// PENDING_ACCOUNT#{accountId} or PENDING_MEETING#{meetingId}.
	PrefixPendingShare   = "PENDINGSHARE#"
	PrefixPendingAccount = "PENDING_ACCOUNT#"
	PrefixPendingMeeting = "PENDING_MEETING#"
)

// PendingShare records an Account- or Meeting-share grant made to an email
// address that has been invited (a Cognito account exists) but has never
// completed a first login -- so no PROFILE row exists yet and
// GetUserByEmail can't resolve it to a userID. AddMember/ShareMeetingByEmail
// write one of these instead of failing outright when the target is in
// exactly this state; DynamoDBRepository.MaterializePendingShares turns it
// into a real AccountMember or Share row the moment GetOrCreateUser creates
// that email's PROFILE row (its first authenticated request after
// accepting the invite), so the grant becomes visible immediately on
// sign-in with no separate "pending invites" step for the invitee.
// PK: PENDINGSHARE#{email}, SK: PENDING_ACCOUNT#{accountId} | PENDING_MEETING#{meetingId}
type PendingShare struct {
	PK              string    `dynamodbav:"PK"`
	SK              string    `dynamodbav:"SK"`
	Email           string    `dynamodbav:"email"`
	Kind            string    `dynamodbav:"kind"` // "account" | "meeting"
	AccountID       string    `dynamodbav:"accountId,omitempty"`
	MeetingID       string    `dynamodbav:"meetingId,omitempty"`
	Role            string    `dynamodbav:"role,omitempty"`       // set when Kind=="account"
	Permission      string    `dynamodbav:"permission,omitempty"` // set when Kind=="meeting"
	InvitedByUserID string    `dynamodbav:"invitedByUserId"`
	InvitedByEmail  string    `dynamodbav:"invitedByEmail,omitempty"`
	CreatedAt       time.Time `dynamodbav:"createdAt"`
	// TTL is a DynamoDB TTL epoch-seconds timestamp (infra/lib/storage-
	// stack.ts's `timeToLiveAttribute: 'ttl'`) -- a mis-typed or stale
	// invite's queued grant is not revocable by its inviter (no list/cancel
	// API, see docs/superpowers/specs/2026-08-04-pending-email-invites-
	// design.md's explicit YAGNI), so this bounds how long it stays
	// claimable instead of granting access at an arbitrary future point.
	TTL        int64  `dynamodbav:"ttl"`
	EntityType string `dynamodbav:"entityType"` // "PENDING_SHARE"
}

// PendingShareTTL is how long a queued grant stays claimable before DynamoDB
// TTL expires it -- long enough that a real invitee who's slow to log in
// isn't punished, short enough that a mis-typed email doesn't sit as a
// silent standing grant forever.
const PendingShareTTL = 30 * 24 * time.Hour

const (
	PendingShareKindAccount = "account"
	PendingShareKindMeeting = "meeting"

	EntityTypePendingShare = "PENDING_SHARE"
)

// SKUserLogin is the sort key for the UserLogin item (see UserLogin doc comment).
const SKUserLogin = "LOGIN"

// Config SK constants
const (
	ConfigSKAllowedDomains = "ALLOWED_DOMAINS"
)

// AllowedDomainsConfig represents the allowed email domains configuration
// PK: CONFIG, SK: ALLOWED_DOMAINS
type AllowedDomainsConfig struct {
	PK         string    `dynamodbav:"PK"`
	SK         string    `dynamodbav:"SK"`
	Domains    []string  `dynamodbav:"domains"`
	UpdatedAt  time.Time `dynamodbav:"updatedAt"`
	UpdatedBy  string    `dynamodbav:"updatedBy"`
	EntityType string    `dynamodbav:"entityType"`
}
